content/start/what_if/index.markdown

```diff
@@ -0,0 +1,321 @@
+
+
+```r
+net <- nn_module(
+  
+  "corr-cnn",
+  
+  initialize = function() {
+    
+    self$conv1 <- nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3)
+    self$conv2 <- nn_conv2d(in_channels = 32, out_channels = 64, kernel_size = 3)
+    self$conv3 <- nn_conv2d(in_channels = 64, out_channels = 128, kernel_size = 3)
+    
+    self$fc1 <- nn_linear(in_features = 14 * 14 * 128, out_features = 128)
+    self$fc2 <- nn_linear(in_features = 128, out_features = 1)
+    
+  },
+  
+  forward = function(x) {
+    
+    x %>% 
+      self$conv1() %>%
+      nnf_relu() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      self$conv2() %>%
+      nnf_relu() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      self$conv3() %>%
+      nnf_relu() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      torch_flatten(start_dim = 2) %>%
+      self$fc1() %>%
+      nnf_relu() %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      self$fc2()
+  }
+)
+```
+
+Batch normalization is less well understood theoretically, but can be extremely effective in some cases. Besides acting as a regularizer, it also stabilizes training and may allow for using higher learning rates.
+
+With batch normalization, our network could look like this:
+
+
+```r
+net <- nn_module(
+  
+  "corr-cnn",
+  
+  initialize = function() {
+    
+    self$conv1 <- nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3)
+    self$conv2 <- nn_conv2d(in_channels = 32, out_channels = 64, kernel_size = 3)
+    self$conv3 <- nn_conv2d(in_channels = 64, out_channels = 128, kernel_size = 3)
+    
+    self$bn1 <- nn_batch_norm2d(num_features = 32)
+    self$bn2 <- nn_batch_norm2d(num_features = 64)
+    self$bn3 <- nn_batch_norm2d(num_features = 128)
+    self$bn4 <- nn_batch_norm1d(num_features = 128)
+    
+    self$fc1 <- nn_linear(in_features = 14 * 14 * 128, out_features = 128)
+    self$fc2 <- nn_linear(in_features = 128, out_features = 1)
+    
+  },
+  
+  forward = function(x) {
+    
+    x %>% 
+      self$conv1() %>%
+      nnf_relu() %>%
+      self$bn1() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      self$conv2() %>%
+      nnf_relu() %>%
+      self$bn2() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      self$conv3() %>%
+      nnf_relu() %>%
+      self$bn3() %>%
+      nnf_avg_pool2d(2) %>%
+      nnf_dropout(p = 0.2) %>%
+      
+      torch_flatten(start_dim = 2) %>%
+      self$fc1() %>%
+      nnf_relu() %>%
+      self$bn4() %>%
+      
+      self$fc2()
+  }
+)
+```
+
+Once you're found that a regularizing measure works -- meaning, performance on the validation set is similar to that on the training set, or maybe even better -- you can go back and add more capacity to the network: add more layers, train for a longer time, etc. Maybe you'll arrive at better performance overall!
```

Comment:
Maybe we could have a 'what if I want to train on the GPU'?

Reply:
yes, you're right, I'll add that! It'll be a frequent question